File: backend/app/api/articles.py
import uuid
from typing import Optional

from fastapi import APIRouter, Depends, Query
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.deps import get_db, get_current_user
from app.models.user import User
from app.schemas.article import (
    ArticleListResponse,
    ArticleResponse,
    ArticleMarkRead,
    ArticleToggleFavorite,
    ArticleToggleReadLater,
)
from app.services.article_service import (
    get_user_articles,
    mark_article_read,
    toggle_favorite,
    toggle_read_later,
)

router = APIRouter(prefix="/articles", tags=["articles"])


@router.get("/", response_model=ArticleListResponse)
async def list_articles(
    unread_only: bool = Query(False),
    limit: int = Query(50, ge=1, le=100),
    offset: int = Query(0, ge=0),
    filter: Optional[str] = Query(None, description="Filter type: 'favorited' or 'read_later'"),
    q: Optional[str] = Query(None, description="Search query (title match)"),
    source_id: Optional[uuid.UUID] = Query(None, description="Filter by source ID"),
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    articles, total = await get_user_articles(
        db, user.id, unread_only, limit, offset,
        filter_type=filter,
        q=q,
        source_id=source_id,
    )
    return ArticleListResponse(
        articles=[ArticleResponse.model_validate(a) for a in articles],
        total=total,
    )


@router.patch("/{article_id}/read")
async def toggle_read(
    article_id: uuid.UUID,
    data: ArticleMarkRead,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    await mark_article_read(db, user.id, article_id, data.is_read)
    return {"ok": True}


@router.patch("/{article_id}/favorite")
async def toggle_article_favorite(
    article_id: uuid.UUID,
    data: ArticleToggleFavorite,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    await toggle_favorite(db, user.id, article_id, data.is_favorited)
    return {"ok": True}


@router.patch("/{article_id}/read-later")
async def toggle_article_read_later(
    article_id: uuid.UUID,
    data: ArticleToggleReadLater,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    await toggle_read_later(db, user.id, article_id, data.is_read_later)
    return {"ok": True}

File: frontend/src/stores/articles.ts
import { defineStore } from 'pinia'
import { ref } from 'vue'
import api from '../api/client'

export interface ArticleSource {
  platform: string
  display_name: string
}

export interface Article {
  id: string
  title: string
  summary: string
  content: string | null
  url: string
  cover_image: string | null
  published_at: string
  source: ArticleSource
  is_read: boolean
  is_favorited: boolean
  is_read_later: boolean
}

export const useArticlesStore = defineStore('articles', () => {
  const articles = ref<Article[]>([])
  const total = ref(0)
  const selected = ref<Article | null>(null)
  const loading = ref(false)

  async function load(unreadOnly = false, limit = 50, offset = 0) {
    loading.value = true
    try {
      const { data } = await api.get('/articles/', { params: { unread_only: unreadOnly, limit, offset } })
      articles.value = data.articles
      total.value = data.total
    } finally {
      loading.value = false
    }
  }

  function select(article: Article) {
    selected.value = article
    if (!article.is_read) {
      markRead(article.id, true)
    }
  }

  async function markRead(id: string, isRead: boolean) {
    await api.patch(`/articles/${id}/read`, { is_read: isRead })
    const a = articles.value.find(a => a.id === id)
    if (a) a.is_read = isRead
  }

  function prependArticle(article: Article) {
    if (!articles.value.find(a => a.id === article.id)) {
      articles.value.unshift(article)
      total.value++
    }
  }

  async function search(q: string) {
    loading.value = true
    try {
      const { data } = await api.get('/articles/', { params: { q } })
      articles.value = data.articles
      total.value = data.total
    } finally {
      loading.value = false
    }
  }

  async function loadFavorites() {
    loading.value = true
    try {
      const { data } = await api.get('/articles/', { params: { filter: 'favorited' } })
      articles.value = data.articles
      total.value = data.total
    } finally {
      loading.value = false
    }
  }

  async function loadReadLater() {
    loading.value = true
    try {
      const { data } = await api.get('/articles/', { params: { filter: 'read_later' } })
      articles.value = data.articles
      total.value = data.total
    } finally {
      loading.value = false
    }
  }

  async function toggleFavorite(id: string) {
    const a = articles.value.find(a => a.id === id)
    const current = a?.is_favorited ?? false
    await api.patch(`/articles/${id}/favorite`, { is_favorited: !current })
    if (a) a.is_favorited = !current
    if (selected.value?.id === id) selected.value.is_favorited = !current
  }

  async function toggleReadLater(id: string) {
    const a = articles.value.find(a => a.id === id)
    const current = a?.is_read_later ?? false
    await api.patch(`/articles/${id}/read-later`, { is_read_later: !current })
    if (a) a.is_read_later = !current
    if (selected.value?.id === id) selected.value.is_read_later = !current
  }

  function selectNext() {
    const idx = articles.value.findIndex(a => a.id === selected.value?.id)
    if (idx < articles.value.length - 1) select(articles.value[idx + 1])
    else if (idx === -1 && articles.value.length > 0) select(articles.value[0])
  }

  function selectPrev() {
    const idx = articles.value.findIndex(a => a.id === selected.value?.id)
    if (idx > 0) select(articles.value[idx - 1])
  }

  return { articles, total, selected, loading, load, select, markRead, prependArticle, search, loadFavorites, loadReadLater, toggleFavorite, toggleReadLater, selectNext, selectPrev }
})
